src/lib/pci_virtio_net_vpnkit.c

```diff
@@ -324,19 +343,56 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
 		init_reply.magic[4],
 		init_reply.version, (int)sizeof(init_reply.commit), init_reply.commit);
 
-	struct msg_ethernet cmd_ethernet = {
-		.command = CMD_ETHERNET,
-	};
-	memcpy(cmd_ethernet.uuid, uuid, sizeof(cmd_ethernet.uuid));
+	if (init_reply.version != init_msg.version) { 
+		fprintf(stderr, "virtio-net-vpnkit: protocol version mismatch: version %d requested, got version %d.\n", 
+			init_msg.version,
+			init_reply.version);
+	}
 
-	if (really_write(fd, (uint8_t*)&cmd_ethernet, sizeof(cmd_ethernet)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to write ethernet cmd\n");
+	if (preferred_ipv4 != 0) {
+		struct msg_preferred_ipv4 cmd_preferred_ipv4 = {
+			.command = CMD_PREFERRED_IPV4,
+			/* VPNKit uses LE, so swap the IP from network byte order */
+			.ip = OSSwapInt32(preferred_ipv4),
+		};
+		memcpy(cmd_preferred_ipv4.uuid, uuid, sizeof(cmd_preferred_ipv4.uuid));
+
+		if (really_write(fd, (uint8_t*)&cmd_preferred_ipv4, sizeof(cmd_preferred_ipv4)) < 0) {
+			fprintf(stderr, "virtio-net-vpnkit: failed to write preferred_ipv4 cmd\n");
+			return -1;
+		}
+	} else {
+		/* No preferred IPv4 address, falling back to requesting a dynamic address */
+		struct msg_ethernet cmd_ethernet = {
+			.command = CMD_ETHERNET,
+			.unused = 0,
+		};
+		memcpy(cmd_ethernet.uuid, uuid, sizeof(cmd_ethernet.uuid));
+
+		if (really_write(fd, (uint8_t*)&cmd_ethernet, sizeof(cmd_ethernet)) < 0) {
+			fprintf(stderr, "virtio-net-vpnkit: failed to write ethernet cmd\n");
+			return -1;
+		}
+	}
+	struct msg_response reply;
+	if (really_read(fd, (uint8_t*)&reply, sizeof(reply)) < 0) {
+		fprintf(stderr, "virtio-net-vpnkit: failed to read response message\n");
 		return -1;
 	}
 
-	if (really_read(fd, (uint8_t*)vif, sizeof(*vif)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to read vif info\n");
-		return -1;
+	struct disconnect_reason *reason;
+
+	switch (reply.response_type) {
+		case RESP_DISCONNECT: 
+			reason = (struct disconnect_reason*)&(reply.data);
+			fprintf(stderr, "virtio-net-vpnkit: server disconnected: %*s\n", reason->len, reason->msg);
+			return -1;
+		case RESP_VIF: 
```

Comment:
I would move this to the top as this is a the normal/common case which indicates success. `RESP_DISCONNECT` and "unknown" are both error cases and should be grouped together